cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/FlowletProgramRunner.java

```diff
@@ -429,34 +430,57 @@ private OutputEmitterFactory outputEmitterFactory(final BasicFlowletContext flow
+          QueueName queueName = null;
+          List<String> consumerFlowlets = Lists.newLinkedList();
           Node flowlet = Node.flowlet(flowletName);
-          for (QueueSpecification queueSpec : Iterables.concat(queueSpecs.row(flowlet).values())) {
-            if (queueSpec.getQueueName().getSimpleName().equals(outputName)
+          Schema schema = schemaGenerator.generate(type.getType());
+          for (Map.Entry<String, Set<QueueSpecification>> entry : queueSpecs.row(flowlet).entrySet()) {
+            for (QueueSpecification queueSpec : entry.getValue()) {
+              if (queueSpec.getQueueName().getSimpleName().equals(outputName)
                 && queueSpec.getOutputSchema().equals(schema)) {
 
-              final MetricsCollector metrics = flowletContext.getProgramMetrics().childCollector(
-                Constants.Metrics.Tag.FLOWLET_QUEUE, queueSpec.getQueueName().getSimpleName());
-              ProducerSupplier producerSupplier = new ProducerSupplier(queueSpec.getQueueName(),
-                                                                       queueClientFactory, new QueueMetrics() {
-                @Override
-                public void emitEnqueue(int count) {
-                  metrics.increment("process.events.out", count);
-                }
-
-                @Override
-                public void emitEnqueueBytes(int bytes) {
-                  // no-op
-                }
-              });
-              producerBuilder.add(producerSupplier);
-              return new DatumOutputEmitter<T>(producerSupplier, schema, datumWriterFactory.create(type, schema));
+                queueName = queueSpec.getQueueName();
+                consumerFlowlets.add(entry.getKey());
+                break;
+              }
             }
           }
+          if (queueName == null) {
+            throw new IllegalArgumentException(String.format("No queue specification found for %s, %s",
+                                                             flowletName, type));
+          }
 
-          throw new IllegalArgumentException(String.format("No queue specification found for %s, %s",
-                                                           flowletName, type));
-
+          // create a metric collector for this queue, and also one for each consumer flowlet
+          final MetricsCollector metrics = flowletContext.getProgramMetrics()
+            .childCollector(Constants.Metrics.Tag.FLOWLET_QUEUE, outputName);
+          final MetricsCollector producerMetrics = metrics.childCollector(
+            Constants.Metrics.Tag.PRODUCER, flowletContext.getFlowletId());
+          final Iterable<MetricsCollector> consumerMetrics =
+            Iterables.transform(consumerFlowlets, new Function<String, MetricsCollector>() {
+              @Nullable
+              @Override
+              public MetricsCollector apply(@Nullable String consumer) {
```

Comment:
Can remove the `@Nullable` as I believe the list that it's transforming won't have null entry. This can eliminate unnecessary warning from IDE (if childCollector() doesn't accept null).